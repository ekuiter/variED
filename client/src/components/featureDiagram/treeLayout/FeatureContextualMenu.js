import React from 'react';
import {DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {ContextualMenu, ContextualMenuItemType} from 'office-ui-fabric-react/lib/ContextualMenu';
import contextualMenuItems from '../../contextualMenuItems';
import {getSetting} from '../../../store/settings';

export const selectMultipleFeaturesContextualMenuItems = (selectedFeatureNames, onSelectAllFeatures, onDeselectAllFeatures, featureModel) => [
    contextualMenuItems.featureDiagram.features.selectAll(onSelectAllFeatures),
    contextualMenuItems.featureDiagram.features.deselectAll(onDeselectAllFeatures),
    {key: 'divider1', itemType: ContextualMenuItemType.Divider},
    contextualMenuItems.featureDiagram.features.newFeatureAbove(selectedFeatureNames, onDeselectAllFeatures, featureModel)
];

export default props => {
    const {
            onDismiss, onSelectAllFeatures, onDeselectAllFeatures,
            isSelectMultipleFeatures, selectedFeatureNames, featureModel, featureName
        } = props,
        {gapSpace} = getSetting(props.settings, 'featureDiagram.treeLayout.overlay');
    const feature = featureModel && featureModel.getFeatureOrDismiss(featureName, featureName, onDismiss);
    if (!feature)
        return null;
    return (
        <ContextualMenu
            target={props.nodeRef}
            onDismiss={onDismiss}
            hidden={!props.featureName}
            isBeakVisible={!isSelectMultipleFeatures}
            gapSpace={isSelectMultipleFeatures ? 2 * gapSpace : gapSpace}
            directionalHint={
                props.direction === 'vertical'
                    ? DirectionalHint.bottomCenter
                    : DirectionalHint.rightCenter}
            items={isSelectMultipleFeatures
                ? selectMultipleFeaturesContextualMenuItems(
                    selectedFeatureNames, onSelectAllFeatures, onDeselectAllFeatures, featureModel)
                : [
                    contextualMenuItems.featureDiagram.feature.new(feature.name, onDismiss),
                    contextualMenuItems.featureDiagram.feature.remove(feature.name, onDismiss),
                    {key: 'divider1', itemType: ContextualMenuItemType.Divider},
                    contextualMenuItems.featureDiagram.feature.rename(feature.name, props.onShowDialog),
                    {key: 'divider2', itemType: ContextualMenuItemType.Divider},
                    contextualMenuItems.featureDiagram.feature.details(feature.name, props.onShowPanel)
                ]
            }/>
    );
}