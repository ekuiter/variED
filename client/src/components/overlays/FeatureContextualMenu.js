import React from 'react';
import {DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {ContextualMenu, ContextualMenuItemType} from 'office-ui-fabric-react/lib/ContextualMenu';
import contextualMenuItems from '../contextualMenuItems';
import {getSetting} from '../../store/settings';
import PropTypes from 'prop-types';
import {FeatureModelType, layoutTypes} from '../../types';
import {LayoutType, SettingsType} from '../../types';

export const selectMultipleFeaturesContextualMenuItems = (selectedFeatureNames, onSelectAllFeatures, onDeselectAllFeatures, featureModel) => [
    contextualMenuItems.featureDiagram.features.selectAll(onSelectAllFeatures),
    contextualMenuItems.featureDiagram.features.deselectAll(onDeselectAllFeatures),
    {key: 'divider1', itemType: ContextualMenuItemType.Divider},
    contextualMenuItems.featureDiagram.features.newFeatureAbove(selectedFeatureNames, onDeselectAllFeatures, featureModel)
];

class FeatureContextualMenu extends React.Component {
    componentDidMount() {
        this.interval = window.setInterval(
            this.forceUpdate.bind(this),
            getSetting(this.props.settings, 'featureDiagram.overlay.throttleUpdate'));
    }

    componentWillUnmount() {
        window.clearInterval(this.interval);
    }

    render() {
        const {
                onDismiss, onSelectAllFeatures, onDeselectAllFeatures,
                isSelectMultipleFeatures, selectedFeatureNames, featureModel, featureName
            } = this.props,
            {gapSpace} = getSetting(this.props.settings, 'featureDiagram.overlay');
        const feature = featureModel && featureModel.getFeatureOrDismiss(featureName, this.props.isOpen, onDismiss);
        if (!feature)
            return null;
        return (
            <ContextualMenu
                target={featureModel.getElement(featureName).querySelector('.rectAndText')}
                onDismiss={onDismiss}
                hidden={!this.props.isOpen}
                isBeakVisible={!isSelectMultipleFeatures}
                gapSpace={isSelectMultipleFeatures ? 2 * gapSpace : gapSpace}
                directionalHint={
                    this.props.featureDiagramLayout === layoutTypes.verticalTree
                        ? DirectionalHint.bottomCenter
                        : DirectionalHint.rightCenter}
                items={isSelectMultipleFeatures
                    ? selectMultipleFeaturesContextualMenuItems(
                        selectedFeatureNames, onSelectAllFeatures, onDeselectAllFeatures, featureModel)
                    : [
                        contextualMenuItems.featureDiagram.feature.new(feature.name, onDismiss),
                        contextualMenuItems.featureDiagram.feature.remove(feature.name, onDismiss),
                        {key: 'divider1', itemType: ContextualMenuItemType.Divider},
                        contextualMenuItems.featureDiagram.feature.rename(feature.name, this.props.onShowOverlay),
                        contextualMenuItems.featureDiagram.feature.setDescription(feature.name, this.props.onShowOverlay),
                        {key: 'divider2', itemType: ContextualMenuItemType.Divider},
                        contextualMenuItems.featureDiagram.feature.details(feature.name, this.props.onShowOverlay)
                    ]
                }/>
        );
    }
}

FeatureContextualMenu.propTypes = {
    onDismiss: PropTypes.func.isRequired,
    onSelectAllFeatures: PropTypes.func.isRequired,
    onDeselectAllFeatures: PropTypes.func.isRequired,
    isSelectMultipleFeatures: PropTypes.bool.isRequired,
    selectedFeatureNames: PropTypes.arrayOf(PropTypes.string).isRequired,
    featureModel: FeatureModelType.isRequired,
    featureName: PropTypes.string.isRequired,
    isOpen: PropTypes.bool.isRequired,
    featureDiagramLayout: LayoutType.isRequired,
    onShowOverlay: PropTypes.func.isRequired,
    settings: SettingsType.isRequired
};

export default FeatureContextualMenu;