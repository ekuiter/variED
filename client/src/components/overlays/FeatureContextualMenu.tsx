/**
 * A Fabric contextual menu for a given feature.
 */

import React from 'react';
import {DirectionalHint} from 'office-ui-fabric-react/lib/Callout';
import {ContextualMenu} from 'office-ui-fabric-react/lib/ContextualMenu';
import commands, {makeDivider} from '../commands';
import {FeatureDiagramLayoutType} from '../../types';
import FeatureComponent, {FeatureComponentProps} from './FeatureComponent';
import {OnShowOverlayFunction, OnCollapseFeaturesFunction, OnCollapseFeaturesBelowFunction, OnExpandFeaturesFunction, OnExpandFeaturesBelowFunction, OnDeselectAllFeaturesFunction, OnRemoveFeaturesFunction, OnAddFeatureAboveFunction, OnAddFeatureBelowFunction, OnRemoveFeaturesBelowFunction, OnSetFeatureAbstractFunction, OnSetFeatureHiddenFunction, OnSetFeatureMandatoryFunction, OnSetFeatureAndFunction, OnSetFeatureOrFunction, OnSetFeatureAlternativeFunction} from '../../store/types';
import {GraphicalFeature} from '../../modeling/types';

type Props = FeatureComponentProps & {
    onDismiss: () => void,
    isOpen: boolean,
    isSelectMultipleFeatures: boolean,
    featureDiagramLayout: FeatureDiagramLayoutType,
    selectedFeatureNames: string[],
    onShowOverlay: OnShowOverlayFunction,
    onCollapseFeatures: OnCollapseFeaturesFunction,
    onCollapseFeaturesBelow: OnCollapseFeaturesBelowFunction,
    onExpandFeatures: OnExpandFeaturesFunction,
    onExpandFeaturesBelow: OnExpandFeaturesBelowFunction,
    onDeselectAllFeatures: OnDeselectAllFeaturesFunction,
    onRemoveFeatures: OnRemoveFeaturesFunction,
    onRemoveFeaturesBelow: OnRemoveFeaturesBelowFunction,
    onAddFeatureBelow: OnAddFeatureBelowFunction,
    onAddFeatureAbove: OnAddFeatureAboveFunction,
    onSetFeatureAbstract: OnSetFeatureAbstractFunction,
    onSetFeatureHidden: OnSetFeatureHiddenFunction,
    onSetFeatureMandatory: OnSetFeatureMandatoryFunction,
    onSetFeatureAnd: OnSetFeatureAndFunction,
    onSetFeatureOr: OnSetFeatureOrFunction,
    onSetFeatureAlternative: OnSetFeatureAlternativeFunction
};

export default class extends FeatureComponent({doUpdate: true})<Props> {
    renderIfFeature(feature: GraphicalFeature) {
        const {
                onDismiss, onDeselectAllFeatures, isSelectMultipleFeatures, selectedFeatureNames, graphicalFeatureModel
            } = this.props,
            {gapSpace} = this.props.settings.featureDiagram.overlay;
        return (
            <ContextualMenu
                target={graphicalFeatureModel!.getElement(feature.name)!.querySelector('.rectAndText')}
                onDismiss={onDismiss}
                hidden={!this.props.isOpen}
                isBeakVisible={!isSelectMultipleFeatures}
                gapSpace={isSelectMultipleFeatures ? 2 * gapSpace : gapSpace}
                directionalHint={
                    this.props.featureDiagramLayout === FeatureDiagramLayoutType.verticalTree
                        ? DirectionalHint.bottomCenter
                        : DirectionalHint.rightCenter}
                items={isSelectMultipleFeatures
                    ? commands.featureDiagram.feature.selectionItems(selectedFeatureNames, onDeselectAllFeatures,
                        this.props.onCollapseFeatures, this.props.onExpandFeatures, this.props.onCollapseFeaturesBelow,
                        this.props.onExpandFeaturesBelow, this.props.onAddFeatureAbove, this.props.onRemoveFeatures,
                        this.props.onRemoveFeaturesBelow, this.props.onSetFeatureAbstract, this.props.onSetFeatureHidden,
                        this.props.onSetFeatureMandatory, this.props.onSetFeatureAnd, this.props.onSetFeatureOr,
                        this.props.onSetFeatureAlternative, graphicalFeatureModel!)
                    : [
                        commands.featureDiagram.feature.newMenu(feature.name, this.props.onAddFeatureBelow, this.props.onAddFeatureAbove, onDismiss),
                        commands.featureDiagram.feature.removeMenu([feature], this.props.onRemoveFeatures, this.props.onRemoveFeaturesBelow, onDismiss),
                        commands.featureDiagram.feature.collapseMenu(
                            [feature], this.props.onCollapseFeatures, this.props.onExpandFeatures,
                            this.props.onCollapseFeaturesBelow, this.props.onExpandFeaturesBelow, onDismiss),
                        makeDivider(),
                        commands.featureDiagram.feature.rename(feature.name, this.props.onShowOverlay),
                        commands.featureDiagram.feature.setDescription(feature.name, this.props.onShowOverlay),
                        commands.featureDiagram.feature.properties([feature], this.props.onSetFeatureAbstract,
                            this.props.onSetFeatureHidden, this.props.onSetFeatureMandatory, this.props.onSetFeatureAnd,
                            this.props.onSetFeatureOr, this.props.onSetFeatureAlternative, onDismiss),
                        makeDivider(),
                        commands.featureDiagram.feature.details(feature.name, this.props.onShowOverlay)
                    ]
                }/>
        );
    }
}
