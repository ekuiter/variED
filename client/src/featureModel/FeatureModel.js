import React from 'react';
import VerticalTreeLayout from './treeLayout/VerticalTreeLayout';
import HorizontalTreeLayout from './treeLayout/HorizontalTreeLayout';

class FeatureModel extends React.Component {
    static defaultProps = {layout: 'verticalTree'};

    static layoutMap = {
        verticalTree: VerticalTreeLayout,
        horizontalTree: HorizontalTreeLayout
    };

    layoutRef = React.createRef();

    key(props) {
        // The key uniquely identifies the layout component instance. If the key changes,
        // the instance is unmounted and a new one is mounted. This is useful for forcing
        return props.debug; // rerenders, i.e. when the debug flag changes.
    }

    render() {
        const {layout, ...props} = this.props,
            LayoutComponent = FeatureModel.layoutMap[layout];

        return (
            <LayoutComponent key={this.key(props)} ref={this.layoutRef} {...props} />
        );
    }

    export() {
        const layout = this.layoutRef.current;
        if (layout && layout.canExport())
            return layout.export().then(console.log);
        else
            return Promise.reject('can not export feature model');
    }
}

export default FeatureModel;