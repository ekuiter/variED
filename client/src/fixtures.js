
export const invalidFeatureModel1 = {};

export const invalidFeatureModel2 = {
    'struct': [],
    'constraints': [],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModel = {
    'struct': [
        {
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'his is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'name': 'JDT'},
                {'type': 'feature', 'name': 'CDT'},
                {'type': 'feature', 'name': 'AJDT'},
                {
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'name': 'CIDE'},
                        {'type': 'feature', 'name': 'FAMILIAR'},
                        {
                            'name': 'FeatureIDE',
                            'type': 'or',
                            'children': [
                                {'type': 'feature', 'name': 'AHEAD'},
                                {'type': 'feature', 'name': 'FeatureHouse'},
                                {'type': 'feature', 'name': 'FeatureCpp'},
                                {'type': 'feature', 'name': 'DeltaJ'},
                                {'type': 'feature', 'name': 'AspectJ'},
                                {'type': 'feature', 'name': 'Munge'},
                                {'type': 'feature', 'name': 'Antenna'},
                                {'type': 'feature', 'name': 'DeltaMontiArc'}
                            ]
                        },
                        {'type': 'feature', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'name': 'MoSoPoLiTe'}
                    ]
                },
                {'type': 'feature', 'name': 'DeltaJEclipsePlugin'}
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'DeltaJ'},
                        {'type': 'var', 'var': 'DeltaJEclipsePlugin'}
                    ]
                }
            ]
        }
    ],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};