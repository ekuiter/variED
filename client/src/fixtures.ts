import {SerializedFeatureModel} from './modeling/types';

/**
 * Some fixtures for use in tests.
 */

export const invalidFeatureModel1 = <SerializedFeatureModel>{};

export const invalidFeatureModel2 = <any>{
    'struct': [],
    'constraints': [],
    'properties': [],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModel = <SerializedFeatureModel>{
    'struct': [
        {
            'ID': 'Eclipse',
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'this is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'ID': 'JDT', 'name': 'JDT'},
                {'type': 'feature', 'ID': 'CDT', 'name': 'CDT'},
                {'type': 'feature', 'ID': 'AJDT', 'name': 'AJDT'},
                {
                    'ID': 'FeatureModeling',
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'ID': 'CIDE', 'name': 'CIDE'},
                        {'type': 'feature', 'ID': 'FAMILIAR', 'name': 'FAMILIAR'},
                        {
                            'ID': 'FeatureIDE',
                            'name': 'FeatureIDE',
                            'description': 'A sample description',
                            'type': 'or',
                            'children': [
                                {'type': 'feature', 'ID': 'AHEAD', 'name': 'AHEAD'},
                                {'type': 'feature', 'ID': 'FeatureHouse', 'name': 'FeatureHouse'},
                                {'type': 'feature', 'ID': 'FeatureCpp', 'name': 'FeatureCpp'},
                                {'type': 'feature', 'ID': 'DeltaJ', 'name': 'DeltaJ'},
                                {'type': 'feature', 'ID': 'AspectJ', 'name': 'AspectJ'},
                                {'type': 'feature', 'ID': 'Munge', 'name': 'Munge'},
                                {'type': 'feature', 'ID': 'Antenna', 'name': 'Antenna'},
                                {'type': 'feature', 'ID': 'DeltaMontiArc', 'name': 'DeltaMontiArc'}
                            ]
                        },
                        {'type': 'feature', 'ID': 'ExtendedFM', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'ID': 'MoSoPoLiTe', 'name': 'MoSoPoLiTe'}
                    ]
                },
                {'type': 'feature', 'ID': 'DeltaJEclipsePlugin', 'name': 'DeltaJEclipsePlugin'}
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'DeltaJ'},
                        {'type': 'var', 'var': 'DeltaJEclipsePlugin'}
                    ]
                }
            ]
        }
    ],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};

export const validFeatureModelWithRemovedFeatures = <SerializedFeatureModel>{
    'struct': [
        {
            'name': 'Eclipse',
            'mandatory': true,
            'description': 'this is a feature which has a description text',
            'type': 'and',
            'children': [
                {'type': 'feature', 'name': 'JDT'},
                {'type': 'feature', 'name': 'CDT'},
                {'type': 'feature', 'name': 'AJDT'},
                {
                    'name': 'FeatureModeling',
                    'type': 'and',
                    'children': [
                        {'type': 'feature', 'name': 'CIDE'},
                        {'type': 'feature', 'name': 'FAMILIAR'},
                        {'type': 'feature', 'name': 'AHEAD'},
                        {'type': 'feature', 'name': 'FeatureHouse'},
                        {'type': 'feature', 'name': 'FeatureCpp'},
                        {'type': 'feature', 'name': 'AspectJ'},
                        {'type': 'feature', 'name': 'Munge'},
                        {'type': 'feature', 'name': 'Antenna'},
                        {'type': 'feature', 'name': 'DeltaMontiArc'},
                        {'type': 'feature', 'name': 'ExtendedFM'},
                        {'type': 'feature', 'name': 'MoSoPoLiTe'}
                    ]
                }
            ]
        }
    ],
    'constraints': [
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {
                            'type': 'disj',
                            'children': [
                                {'type': 'var', 'var': 'AHEAD'},
                                {
                                    'type': 'disj',
                                    'children': [
                                        {'type': 'var', 'var': 'FeatureHouse'},
                                        {
                                            'type': 'disj',
                                            'children': [
                                                {'type': 'var', 'var': 'Munge'},
                                                {'type': 'var', 'var': 'Antenna'}
                                            ]
                                        }
                                    ]
                                }
                            ]
                        },
                        {'type': 'var', 'var': 'JDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'FeatureCpp'},
                        {'type': 'var', 'var': 'CDT'}
                    ]
                }
            ]
        },
        {
            'type': 'rule',
            'children': [
                {
                    'type': 'imp',
                    'children': [
                        {'type': 'var', 'var': 'AspectJ'},
                        {'type': 'var', 'var': 'AJDT'}
                    ]
                }
            ]
        }
    ],
    'calculations': {
        'Auto': true,
        'Features': true,
        'Constraints': true,
        'Redundant': true,
        'Tautology': true
    },
    'comments': [],
    'featureOrder': {'userDefined': false}
};