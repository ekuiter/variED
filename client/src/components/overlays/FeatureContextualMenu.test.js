import React from 'react';
import {shallow} from 'enzyme';
import FeatureContextualMenu from './FeatureContextualMenu';
import {layoutTypes} from '../../types';
import {defaultSettings} from '../../store/settings';
import FeatureModel from '../../server/FeatureModel';
import {validFeatureModel} from '../../fixtures';
import {ContextualMenu} from 'office-ui-fabric-react/lib/ContextualMenu';
import {DirectionalHint} from 'office-ui-fabric-react/lib/Callout';

describe('FeatureContextualMenu', () => {
    const featureContextualMenu = (selectedFeatureNames = []) => {
        const featureModel = new FeatureModel(validFeatureModel, []),
            mock = jest.fn();
        featureModel.getElement = jest.fn().mockReturnValue({
            querySelector: () => '<target element>'
        });
        return shallow(
            <FeatureContextualMenu
                isOpen={true}
                onDismiss={mock}
                featureDiagramLayout={layoutTypes.verticalTree}
                settings={defaultSettings}
                onShowOverlay={mock}
                onDeselectAllFeatures={mock}
                onCollapseFeature={mock}
                onExpandFeature={mock}
                featureModel={featureModel}
                isSelectMultipleFeatures={selectedFeatureNames.length > 0}
                selectedFeatureNames={selectedFeatureNames}
                featureName="FeatureIDE"/>
        );
    };

    it('renders correctly for a single feature', () => {
        const wrapper = featureContextualMenu();
        expect(wrapper.prop('target')).toBe('<target element>');
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly for a selection of features', () => {
        const wrapper = featureContextualMenu(['FeatureIDE', 'Eclipse']);
        expect(wrapper.prop('target')).toBe('<target element>');
        expect(wrapper).toMatchSnapshot();
    });

    it('switches direction depending on the layout', () => {
        const wrapper = featureContextualMenu();
        expect(wrapper.find(ContextualMenu).prop('directionalHint')).toBe(DirectionalHint.bottomCenter);
        wrapper.setProps({featureDiagramLayout: layoutTypes.horizontalTree});
        expect(wrapper.find(ContextualMenu).prop('directionalHint')).toBe(DirectionalHint.rightCenter);
    });
});