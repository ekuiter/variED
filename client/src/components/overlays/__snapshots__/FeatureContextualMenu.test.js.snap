// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureContextualMenu renders correctly for a selection of features 1`] = `
<StyledWithResponsiveMode
  directionalHint={5}
  gapSpace={10}
  hidden={false}
  isBeakVisible={false}
  items={
    Array [
      Object {
        "disabled": true,
        "iconProps": Object {
          "iconName": "Add",
        },
        "key": "newAbove",
        "onClick": [Function],
        "text": "New feature above",
      },
      Object {
        "disabled": true,
        "iconOnly": false,
        "iconProps": Object {
          "iconName": "Remove",
        },
        "key": "removeMenu",
        "onClick": [Function],
        "split": true,
        "subMenuProps": Object {
          "items": Array [
            Object {
              "iconProps": Object {
                "iconName": "Remove",
              },
              "key": "remove",
              "onClick": [Function],
              "secondaryText": "⌫",
              "text": "Remove features",
            },
            Object {
              "disabled": true,
              "iconProps": Object {
                "iconName": "Remove",
              },
              "key": "removeBelow",
              "onClick": [Function],
              "text": "Remove features below",
            },
          ],
        },
        "text": "Remove",
      },
    ]
  }
  onDismiss={[Function]}
  target="<target element>"
/>
`;

exports[`FeatureContextualMenu renders correctly for a single feature 1`] = `
<StyledWithResponsiveMode
  directionalHint={5}
  gapSpace={5}
  hidden={false}
  isBeakVisible={true}
  items={
    Array [
      Object {
        "iconOnly": false,
        "iconProps": Object {
          "iconName": "Add",
        },
        "key": "newMenu",
        "onClick": [Function],
        "split": true,
        "subMenuProps": Object {
          "items": Array [
            Object {
              "iconProps": Object {
                "iconName": "Add",
              },
              "key": "newBelow",
              "onClick": [Function],
              "secondaryText": "Ctrl+N",
              "text": "New feature below",
            },
            Object {
              "disabled": false,
              "iconProps": Object {
                "iconName": "Add",
              },
              "key": "newAbove",
              "onClick": [Function],
              "text": "New feature above",
            },
          ],
        },
        "text": "New",
      },
      Object {
        "disabled": false,
        "iconOnly": false,
        "iconProps": Object {
          "iconName": "Remove",
        },
        "key": "removeMenu",
        "onClick": [Function],
        "split": true,
        "subMenuProps": Object {
          "items": Array [
            Object {
              "iconProps": Object {
                "iconName": "Remove",
              },
              "key": "remove",
              "onClick": [Function],
              "secondaryText": "⌫",
              "text": "Remove feature",
            },
            Object {
              "disabled": false,
              "iconProps": Object {
                "iconName": "Remove",
              },
              "key": "removeBelow",
              "onClick": [Function],
              "text": "Remove features below",
            },
          ],
        },
        "text": "Remove",
      },
      Object {
        "disabled": false,
        "iconOnly": false,
        "iconProps": Object {
          "iconName": "CollapseContentSingle",
        },
        "key": "collapseMenu",
        "onClick": [Function],
        "split": true,
        "subMenuProps": Object {
          "items": Array [
            Object {
              "iconProps": Object {
                "iconName": "CollapseContentSingle",
              },
              "key": "collapse",
              "onClick": [Function],
              "secondaryText": "Ctrl+C",
              "text": "Collapse feature",
            },
            Object {
              "iconProps": Object {
                "iconName": "CollapseContent",
              },
              "key": "collapseBelow",
              "onClick": [Function],
              "text": "Collapse features below",
            },
            Object {
              "iconProps": Object {
                "iconName": "ExploreContent",
              },
              "key": "expandBelow",
              "onClick": [Function],
              "text": "Expand features below",
            },
          ],
        },
        "text": "Collapse",
      },
      Object {
        "itemType": 1,
        "key": "divider",
      },
      Object {
        "iconProps": Object {
          "iconName": "Rename",
        },
        "key": "rename",
        "onClick": [Function],
        "secondaryText": "F2",
        "text": "Rename…",
      },
      Object {
        "iconProps": Object {
          "iconName": "TextDocument",
        },
        "key": "setDescription",
        "onClick": [Function],
        "text": "Set description…",
      },
      Object {
        "iconProps": Object {
          "iconName": "FieldNotChanged",
        },
        "key": "propertiesMenu",
        "subMenuProps": Object {
          "items": Array [
            Object {
              "canCheck": true,
              "isChecked": false,
              "key": "abstract",
              "onClick": [Function],
              "text": "Abstract",
            },
            Object {
              "canCheck": true,
              "isChecked": true,
              "key": "concrete",
              "onClick": [Function],
              "text": "Concrete",
            },
            Object {
              "itemType": 1,
              "key": "divider",
            },
            Object {
              "canCheck": true,
              "isChecked": false,
              "key": "hidden",
              "onClick": [Function],
              "text": "Hidden",
            },
            Object {
              "itemType": 1,
              "key": "divider",
            },
            Object {
              "canCheck": true,
              "disabled": false,
              "isChecked": false,
              "key": "mandatory",
              "onClick": [Function],
              "text": "Mandatory",
            },
            Object {
              "canCheck": true,
              "disabled": false,
              "isChecked": true,
              "key": "optional",
              "onClick": [Function],
              "text": "Optional",
            },
            Object {
              "itemType": 1,
              "key": "divider",
            },
            Object {
              "canCheck": true,
              "disabled": false,
              "isChecked": false,
              "key": "and",
              "onClick": [Function],
              "text": "And",
            },
            Object {
              "canCheck": true,
              "disabled": false,
              "isChecked": true,
              "key": "or",
              "onClick": [Function],
              "text": "Or",
            },
            Object {
              "canCheck": true,
              "disabled": false,
              "isChecked": false,
              "key": "alternative",
              "onClick": [Function],
              "text": "Alternative",
            },
          ],
        },
        "text": "Properties",
      },
      Object {
        "itemType": 1,
        "key": "divider",
      },
      Object {
        "iconOnly": true,
        "iconProps": Object {
          "iconName": "Info",
        },
        "key": "details",
        "onClick": [Function],
        "secondaryText": "Ctrl+,",
        "text": "Details…",
      },
    ]
  }
  onDismiss={[Function]}
  target="<target element>"
/>
`;
